STORM-2896: Add tool to help users migrate offsets from storm-kafka t…

srdo opened the pull request:
…o storm-kafka-client

I'm not sure if we want this in master as well? 

I'd like to make a similar tool for the Trident spouts, but those aren't covered in this PR. I tested this by running the storm-perf Kafka null bolt topology (with and without wildcard topics) and migrating the resulting offsets into Kafka.

srdo commented:
Added tool to migrate Trident spouts.

HeartSaVioR commented:
I think it would be last call to migrate if this is shipped to master branch. So I think it would be good to have in 2.0.0. We may not want to have it in 3.0.0 version line.

srdo commented:
Thanks, will make a master branch version.

HeartSaVioR commented on external/storm-kafka-migration/src/main/java/org/apache/storm/kafka/migration/KafkaTridentSpoutMigration.java:
```diff
@@ -69,24 +70,40 @@ public String toString() {
+    /**
+     * Get value for key. Error if value is null or not the expected type.
+     */
+    public static <T> T getOrError(Map<String, Object> conf, String key) {
```
Not a big deal (because the storm-kafka-migration module is just a utility), but getOrError() looks like better to place to separate class to decouple two independent classes.  Just a 2 cents and I'm OK even we keep it as it is.

HeartSaVioR commented on external/storm-kafka-migration/src/main/java/org/apache/storm/kafka/migration/KafkaSpoutMigration.java:
```diff
@@ -54,25 +55,30 @@
     public static void main(String[] args) throws Exception {
         if (args.length != 1) {
             System.err.println("Args: confFile");
+            return;
```
You may want to exit with explicit return code: it will help scripts which leverages this module. Same for trident migration.

srdo commented:
@HeartSaVioR Thanks for reviewing. Addressed the comments. Yes, I migrated offsets from both a regular spout and a trident spout, both for a regular topic and a wildcard subscription. For the regular spout, I think it's good enough to show that the offsets a migrated to the specified consumer group, so I didn't bother setting up a storm-kafka-client spout. For Trident I set up a storm-kafka-client Trident topology and checked that the new topology picked up where the storm-kafka topology left off.

HeartSaVioR commented on docs/storm-kafka.md:
```diff
@@ -4,6 +4,10 @@ layout: documentation
 documentation: true
 ---
 
+## Deprecated
```
+1 on adding explanation of deprecation.
